drivers/input/input_pinnacle.c

```diff
@@ -0,0 +1,627 @@
+#define PINNACLE_REG_PACKET_BYTE1     (0x13) /* R */
+#define PINNACLE_REG_PACKET_BYTE2     (0x14) /* R */
+#define PINNACLE_REG_PACKET_BYTE3     (0x15) /* R */
+#define PINNACLE_REG_PACKET_BYTE4     (0x16) /* R */
+#define PINNACLE_REG_PACKET_BYTE5     (0x17) /* R */
+#define PINNACLE_REG_GPIO_A_CTRL      (0x18) /* R/W */
+#define PINNACLE_REG_GPIO_A_DATA      (0x19) /* R/W */
+#define PINNACLE_REG_GPIO_B_CTRL_DATA (0x1A) /* R/W */
+#define PINNACLE_REG_ERA_VALUE        (0x1B) /* R/W */
+#define PINNACLE_REG_ERA_ADDR_HIGH    (0x1C) /* R/W */
+#define PINNACLE_REG_ERA_ADDR_LOW     (0x1D) /* R/W */
+#define PINNACLE_REG_ERA_ADDR_CTRL    (0x1E) /* R/W */
+#define PINNACLE_REG_PRODUCT_ID       (0x1F) /* R */
+
+/* Firmware ASIC ID value */
+#define PINNACLE_FIRMWARE_ID (0x07)
+
+/* Status1 definition */
+#define PINNACLE_STATUS1_SW_DR BIT(2)
+#define PINNACLE_STATUS1_SW_CC BIT(3)
+
+/* SysConfig1 definition */
+#define PINNACLE_SYS_CONFIG1_RESET          BIT(0)
+#define PINNACLE_SYS_CONFIG1_SHUTDOWN       BIT(1)
+#define PINNACLE_SYS_CONFIG1_LOW_POWER_MODE BIT(2)
+
+/* FeedConfig1 definition */
+#define PINNACLE_FEED_CONFIG1_FEED_ENABLE        BIT(0)
+#define PINNACLE_FEED_CONFIG1_DATA_MODE_ABSOLUTE BIT(1)
+#define PINNACLE_FEED_CONFIG1_FILTER_DISABLE     BIT(2)
+#define PINNACLE_FEED_CONFIG1_X_DISABLE          BIT(3)
+#define PINNACLE_FEED_CONFIG1_Y_DISABLE          BIT(4)
+/* X max to 0 */
+#define PINNACLE_FEED_CONFIG1_X_DATA_INVERT      BIT(6)
+/* Y max to 0 */
+#define PINNACLE_FEED_CONFIG1_Y_DATA_INVERT      BIT(7)
+
+/* FeedConfig2 definition */
+#define PINNACLE_FEED_CONFIG2_INTELLIMOUSE_ENABLE   BIT(0)
+#define PINNACLE_FEED_CONFIG2_ALL_TAPS_DISABLE      BIT(1)
+#define PINNACLE_FEED_CONFIG2_SECONDARY_TAP_DISABLE BIT(2)
+#define PINNACLE_FEED_CONFIG2_SCROLL_DISABLE        BIT(3)
+#define PINNACLE_FEED_CONFIG2_GLIDE_EXTEND_DISABLE  BIT(4)
+/* 90 degrees rotation */
+#define PINNACLE_FEED_CONFIG2_SWAP_X_AND_Y          BIT(7)
+
+/* Special definitions */
+#define PINNACLE_FB (0xFB) /* Filler byte */
+#define PINNACLE_FC (0xFC) /* Auto-increment byte */
+
+/* Read and write masks */
+#define PINNACLE_READ_MSK  (0xA0)
+#define PINNACLE_WRITE_MSK (0x80)
+
+/* Read and write register addresses */
+#define PINNACLE_READ_REG(addr)  (PINNACLE_READ_MSK | addr)
+#define PINNACLE_WRITE_REG(addr) (PINNACLE_WRITE_MSK | addr)
+
+struct pinnacle_active_range {
+	uint16_t x_min;
+	uint16_t x_max;
+	uint16_t y_min;
+	uint16_t y_max;
+};
+
+struct pinnacle_resolution {
+	uint16_t x;
+	uint16_t y;
+};
+
+struct pinnacle_bus;
+
+typedef bool (*pinnacle_bus_is_ready)(const struct pinnacle_bus *bus);
+typedef int (*pinnacle_bus_write)(const struct pinnacle_bus *bus, uint8_t address, uint8_t value);
+typedef int (*pinnacle_bus_read)(const struct pinnacle_bus *bus, uint8_t address, uint8_t *value);
+typedef int (*pinnacle_bus_seq_read)(const struct pinnacle_bus *bus, uint8_t address, uint8_t *data,
+				     uint8_t count);
```

Comment:
how would you feel about folding this into the struct directly? don't quite see the point of doing the typedef to use it only once, it's not like it serves any other purpose, I think it may let you drop the struct forward declaration as well